modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/badrows/KinesisSink.scala

```diff
@@ -0,0 +1,248 @@
+          Success(result.nextBatchAttempt.toList)
+        }
+      }
+
+  private def executeKinesisRequest(requestEntires: List[PutRecordsRequestEntry]): Try[TryBatchResult] = {
+    val putRecordsRequest =
+      new PutRecordsRequest()
+        .withStreamName(config.streamName)
+        .withRecords(requestEntires.asJava)
+
+    Try(writeDataToKinesis(putRecordsRequest))
+      .map(TryBatchResult.build(requestEntires, _))
+  }
+
+  private def buildRecords(keyedData: List[KeyedData]): List[PutRecordsRequestEntry] =
+    // Data was prepended (when building batch) to 'keyedData' list, which results in reversed original order of items.
+    // We have to 'reverse' it now to bring it back.
+    keyedData.reverse.map { data =>
+      new PutRecordsRequestEntry()
+        .withPartitionKey(data.key)
+        .withData(ByteBuffer.wrap(data.content))
+    }
+
+  private def failureMessageForInternalFailures(records: List[PutRecordsRequestEntry], result: TryBatchResult) = {
+    val exampleMessage = result.exampleInternalError.getOrElse("none")
+    s"Writing ${records.size} records to '${config.streamName}' errored with internal failures. Example error message [$exampleMessage]."
+  }
+
+  private def failureMessageForThrottling(records: List[PutRecordsRequestEntry]): String =
+    s"Exceeded Kinesis provisioned throughput: ${records.size} records failed writing to '${config.streamName}'."
+}
+
+object KinesisSink {
+  type WriteDataToKinesis = PutRecordsRequest => PutRecordsResult
+  type FailedWriteRecords = List[PutRecordsRequestEntry]
+
+  def createFrom(config: Config.Output.BadSink.Kinesis): KinesisSink = {
+    val client = Cloud.createKinesisClient(config.region)
+    val writeDataToKinesis: WriteDataToKinesis = client.putRecords
+    new KinesisSink(writeDataToKinesis, config)
+  }
+
+  final case class TryBatchResult(
+    nextBatchAttempt: Vector[PutRecordsRequestEntry],
+    hadSuccess: Boolean,
+    wasThrottled: Boolean,
+    exampleInternalError: Option[String]
+  ) {
+    // Only retry the exact same again if no record was successfully inserted, and all the errors
+    // were not throughput exceeded exceptions
+    def shouldRetrySameBatch: Boolean =
+      !hadSuccess && !wasThrottled
+  }
+
+  object TryBatchResult {
+
+    implicit def tryBatchResultMonoid: Monoid[TryBatchResult] =
+      new Monoid[TryBatchResult] {
+        override val empty: TryBatchResult = TryBatchResult(Vector.empty, hadSuccess = false, wasThrottled = false, None)
+        override def combine(x: TryBatchResult, y: TryBatchResult): TryBatchResult =
+          TryBatchResult(
+            x.nextBatchAttempt ++ y.nextBatchAttempt,
+            x.hadSuccess || y.hadSuccess,
+            x.wasThrottled || y.wasThrottled,
+            x.exampleInternalError.orElse(y.exampleInternalError)
+          )
+      }
+
+    def build(records: List[PutRecordsRequestEntry], result: PutRecordsResult): TryBatchResult =
+      if (result.getFailedRecordCount.toInt =!= 0)
```

Comment:
Why `=!=` instead of simple `!=`? So if didn't use int here, but different incompatible type like string:

* `=!=` wouldn't compile
* `!=` would give us a warning

So I believe it's about compile time safety. BTW copied from [enrich](https://github.com/snowplow/enrich/blob/34e5b7945e689b9318215ab3d75760e91fac06e3/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L254) :D